server/routerlicious/packages/lambdas/src/scribe/summaryWriter.ts

```diff
@@ -702,79 +623,154 @@ export class SummaryWriter implements ISummaryWriter {
 	): Promise<ISequencedDocumentMessage[]> {
 		if (lt - gt <= 1) {
 			return [];
-		} else if (pending.length > 0) {
-			// Check for logtail ops in the unprocessed ops currently present in memory
-			const pendingOps = pending
+		}
+
+		// Define these for the finally block, these should be used as const
+		let logTailFromLastSummary: ISequencedDocumentMessage[] = [];
+		let logTailFromPending: ISequencedDocumentMessage[] = [];
+		let logtailFromMemory: ISequencedDocumentMessage[] = [];
+		let logtailGaps: number[][] = [];
+		let retrievedGaps: ISequencedDocumentMessage[][] = [];
+		let finalLogTail: ISequencedDocumentMessage[] = [];
+
+		try {
+			// Read from last summary logtail first, which is in memory
+			logTailFromLastSummary =
+				this.lastSummaryMessages?.filter(
+					(ms) => ms.sequenceNumber > gt && ms.sequenceNumber < lt,
+				) ?? [];
+
+			logTailFromPending = pending
 				.filter(
 					(op) => op.operation.sequenceNumber > gt && op.operation.sequenceNumber < lt,
 				)
 				.map((op) => op.operation);
-			if (
-				pendingOps.length > 0 &&
-				pendingOps[0].sequenceNumber === gt + 1 &&
-				pendingOps[pendingOps.length - 1].sequenceNumber === lt - 1
-			) {
-				Lumberjack.info(
-					`LogTail of length ${pendingOps.length} between ${gt + 1} and ${
-						lt - 1
-					} fetched from pending ops`,
-					this.lumberProperties,
-				);
-				return pendingOps;
+
+			logtailFromMemory = dedupeSortedArray(
+				mergeArrays(logTailFromLastSummary, logTailFromPending, (op) => op.sequenceNumber),
+				(op) => op.sequenceNumber,
+			);
+
+			logtailGaps = this.findMissingGapsInLogtail(logtailFromMemory, gt + 1, lt - 1);
+			if (logtailGaps.length === 0) {
+				finalLogTail = logtailFromMemory;
+				return finalLogTail;
 			}
-		}
 
-		let logTail: ISequencedDocumentMessage[] = [];
+			retrievedGaps = await Promise.all(
+				logtailGaps.map(async (gap) => {
+					return this.retrieveOps(gap[0] - 1, gap[1] + 1);
```

Comment:
Just educate myself, where does this gap comes from? Why do gap[0] - 1, gap[1] + 1 here? 
Maybe add the comments for inclusive or exclusive?

Reply:
Refactored to use named variables